documents/system-design-document.md

```diff
@@ -0,0 +1,455 @@
+# System Design Document for TODO List Web App
+
+## 1. Overview
+  The objective of this project is to develop a user-friendly TODO list web application that empowers users to efficiently manage their tasks. The application will provide core functionalities such as creating, reading, updating, deleting, searching, and filtering tasks within a secure environment.
+
+  For detailed project requirements, please refer to the [Product Requirements Document (PRD)](https://github.com/what-a-cool-team/docs/blob/main/prd/what-a-todo.md).
+
+## 2. Functional Requirements
+| Priority | Requirements |
+|----------|--------------|
+| P0 | Users are able to create an account |
+| P0 | Users are able to delete their account |
+| P0 | Users are able to log in securely |
+| P0 | Users are able create new tasks with a name, and optionally add description and tags |
+| P0 | Users are able to update existing tasks to change details such as name, description, status, and tags|
+| P1 | Users are able to view a list of all their tasks |
+| P2 | Users are able to delete tasks they no longer need |
+| P2 | Users are able to search tasks by name or tags |
+| P2 | Users are able to filter tasks by status |
+
+## 3. Operational requirements
+  * **Security**: 
+    * The system should grant users access only to the tasks that they own.
+    * The system should allow users to log in and out using their registered ID and password.
+    * The system should allow users to delete their accounts.
+    * Administrators should be restricted from accessing user passwords.
+    
+## 4. Domain Model Design
+
+### 4.1 Entities
+|Entity| ID example |
+|------|--------------|
+|User|
+|Task|
+|Tag|
+|Status| "Created", "Committed", "In Progress", "Completed" |
+
+
+### 4.2 Attributes
+
+| Entity | Question | Logical type | Example value | 
+| -------- | ------- | ---------------| ------------- | 
+|User| What is the account ID of this user? | string  | "user"| 
+|Task| What is the name of this task? | string | "Grocery shopping" | 
+|Task| What is the description for this task?" | string | "1. Meat,<br> 2. Onion,"| 
+| Task | What is the status of this task? | enum | "In Progress" |
+| Tag | What is the name of this tag? | string | "Daily" |
+
+### 4.3 Links
+|Entity * Entity| Sentences | Cardinality |
+|---------------|-----------|-------------|
+|User < Task|User creates many Tasks<br> Task is created by *only one* User|1:N|
+|Task = Tag |Task can have *several* tags assigned<br> Tag can be assigned to *several* tasks |M:N |
+
+### Domain Model Diagram
+![alt text](diagram.png)
+
+## 5. Tables
+ 
+```sql
+CREATE TABLE users (
+  id SERIAL PRIMARY KEY,
+  account_id VARCHAR(64) NOT NULL
+);
+
+CREATE TABLE tasks (
+  id SERIAL PRIMARY KEY,
```

Comment:
Didn't we say we don't want duplicated ids anymore? Just one unique id type is enough